tests/test_util_modules.py

```diff
@@ -1,129 +1,299 @@
+    assert future.cancel() is False
+    assert future.done() and not future.running() and not future.cancelled()
 
-    f2.set_result(['abacaba', 123])
-    assert f1.result() == ['abacaba', 123]
+    future = hivemind.MPFuture()
+    with pytest.raises(concurrent.futures.TimeoutError):
+        future.result(timeout=1e-3)
+
+    future.set_result(['abacaba', 123])
+    assert future.result() == ['abacaba', 123]
 
 
 def test_mpfuture_exception():
-    f1, f2 = hivemind.MPFuture.make_pair()
-    with pytest.raises(TimeoutError):
-        f1.exception(timeout=1e-3)
+    future = hivemind.MPFuture()
+    with pytest.raises(concurrent.futures.TimeoutError):
+        future.exception(timeout=1e-3)
 
-    f2.set_exception(NotImplementedError())
+    def _proc(future):
+        future.set_exception(NotImplementedError())
 
-    for future in [f1, f2]:
-        assert isinstance(future.exception(), NotImplementedError)
-        with pytest.raises(NotImplementedError):
-            future.result()
-        assert future.cancel() is False
-        assert future.done() and not future.running() and not future.cancelled()
+    p = mp.Process(target=_proc, args=(future,))
+    p.start()
+    p.join()
+
+    assert isinstance(future.exception(), NotImplementedError)
+    with pytest.raises(NotImplementedError):
+        future.result()
+    assert future.cancel() is False
+    assert future.done() and not future.running() and not future.cancelled()
 
 
 def test_mpfuture_cancel():
-    f1, f2 = hivemind.MPFuture.make_pair()
-    assert not f2.cancelled()
-    f1.cancel()
-    for future in [f1, f2]:
-        with pytest.raises(CancelledError):
+    future = hivemind.MPFuture()
+    assert not future.cancelled()
+    future.cancel()
+    evt = mp.Event()
+
+    def _proc():
+        with pytest.raises(concurrent.futures.CancelledError):
             future.result()
-        with pytest.raises(CancelledError):
+        with pytest.raises(concurrent.futures.CancelledError):
             future.exception()
-        with pytest.raises(FutureStateError):
+        with pytest.raises(InvalidStateError):
             future.set_result(123)
-        with pytest.raises(FutureStateError):
+        with pytest.raises(InvalidStateError):
             future.set_exception(NotImplementedError())
         assert future.cancelled() and future.done() and not future.running()
+        evt.set()
+
+    p = mp.Process(target=_proc)
+    p.start()
+    p.join()
+    assert evt.is_set()
 
 
 def test_mpfuture_status():
-    f1, f2 = hivemind.MPFuture.make_pair()
-    assert f1.set_running_or_notify_cancel() is True
-    for future in [f1, f2]:
-        assert future.running() and not future.done() and not future.cancelled()
-        with pytest.raises(RuntimeError):
-            future.set_running_or_notify_cancel()
-    f2.cancel()
-    for future in [f1, f2]:
+    future = hivemind.MPFuture()
+
+    def _proc1(future):
+        assert future.set_running_or_notify_cancel() is True
```

Comment:
If this assert fails, we won't know about it.

Reply:
we do, in the logs